### PR TITLE
[SPARK-35177][SQL] Fix arithmetic overflow in parsing the minimal interval by `IntervalUtils.fromYearMonthString`

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
IntervalUtils.fromYearMonthString should handle Int.MinValue months correctly.
In current logic, just use `Math.addExact(Math.multiplyExact(years, 12), months)` to calculate  negative total months will overflow when actual total months is Int.MinValue, this pr fixes this bug.

### Why are the changes needed?
IntervalUtils.fromYearMonthString should handle Int.MinValue months correctly

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @MaxGekk

**Comment by sarutak** [comment]
@AngersZhuuuu JIRA ID seems to be wrong. Could you confirm it?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu JIRA ID seems to be wrong. Could you confirm it?

Done.

**Comment by AngersZhuuuu** [comment]
retest this please

**Review by MaxGekk — COMMENTED** [review]
Could you make PR's title more precise, for example:
[SPARK-35177][SQL] Fix arithmetic overflow in parsing the minimal interval by `IntervalUtils.fromYearMonthString`

**Comment by AngersZhuuuu** [comment]
> Could you make PR's title more precise, for example:
> [[SPARK-35177](https://issues.apache.org/jira/browse/SPARK-35177)][SQL] Fix arithmetic overflow in parsing the minimal interval by `IntervalUtils.fromYearMonthString`

Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala [hunk 1]
```diff
@@ -169,6 +169,19 @@ class IntervalUtilsSuite extends SparkFunSuite with SQLHelper {
       fromYearMonthString)
     failFuncWithInvalidInput("-\t99-15", "Interval string does not match year-month format",
       fromYearMonthString)
+
+    assert(fromYearMonthString("178956970-6") == new CalendarInterval(2147483646, 0, 0))
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala:173]: nit: let's use `Int.MaxValue - 1`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala [hunk 2]
```diff
@@ -169,6 +169,19 @@ class IntervalUtilsSuite extends SparkFunSuite with SQLHelper {
...
+    assert(fromYearMonthString("178956970-7") == new CalendarInterval(2147483647, 0, 0))
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala:174]: Please, replace 2147483647 by Int.MaxValue

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala:174]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala:173]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 3]
```diff
@@ -100,11 +100,15 @@ object IntervalUtils {
    */
   def fromYearMonthString(input: String): CalendarInterval = {
     require(input != null, "Interval year-month string must be not null")
-    def toInterval(yearStr: String, monthStr: String): CalendarInterval = {
+    def toInterval(yearStr: String, monthStr: String, negative: Boolean): CalendarInterval = {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:103]: Let's re-use range check in toLongWithRange and simplify the code:
```scala
    def toInterval(yearStr: String, monthStr: String, sign: Int): CalendarInterval = {
      try {
        val years = toLongWithRange(YEAR, yearStr, 0, Integer.MAX_VALUE / MONTHS_PER_YEAR)
        val totalMonths = sign * (years * MONTHS_PER_YEAR + toLongWithRange(MONTH, monthStr, 0, 11))
        new CalendarInterval(Math.toIntExact(totalMonths), 0, 0)
      } catch {
        case NonFatal(e) =>
          throw new IllegalArgumentException(
            s"Error parsing interval year-month string: ${e.getMessage}", e)
      }
    }
```

**Review by MaxGekk — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: DOne

**Comment by MaxGekk** [comment]
+1, LGTM. GA passed. Merging to master.
Screenshot 2021-04-22 at 11 56 38
Thank you, @AngersZhuuuu .

**Pull request closed without merging**